rust/src/io/reader.rs

```diff
@@ -60,7 +77,196 @@ pub async fn read_manifest(object_store: &ObjectStore, path: &Path) -> Result<Ma
+        self.projection.as_ref().unwrap()
+    }
+
+    /// Read a batch of data from the file.
+    ///
+    /// The schema of the returned [RecordBatch] is set by [`FileReader::schema()`].
+    pub async fn read_batch(&self, batch_id: i32) -> Result<RecordBatch> {
+        let schema = self.projection.as_ref().unwrap();
+        // TODO spawn more threads
+        let mut arrs = vec![];
+        for field in schema.fields.iter() {
+            let arr = self.read_array(&field, batch_id).await?;
+            println!("Read array: {:?}\n", arr);
+            arrs.push(arr);
+        }
+        RecordBatch::try_new(Arc::new(schema.into()), arrs)
+            .map_err(|e| Error::new(ErrorKind::Other, e.to_string()))
+    }
+
+    /// Read primitive array for batch `batch_idx`.
+    ///
+    async fn read_primitive_array(&self, field: &Field, batch_id: i32) -> Result<ArrayRef> {
+        let column = field.id;
+        let page_info = self.page_table.get(column, batch_id).ok_or(Error::new(
+            ErrorKind::InvalidData,
+            format!(
+                "field: {} batch {} does not exist in page table",
+                field, batch_id
+            ),
+        ))?;
+        use DataType::*;
+
+        macro_rules! create_plain_decoder {
+            ($a:ty) => {
+                Box::new(PlainDecoder::<$a>::new(
+                    &self.object_reader,
+                    page_info.position,
+                    page_info.length,
+                )?)
+            };
+        }
+
+        let decoder: Box<dyn Decoder> = match field.data_type() {
+            Int8 => create_plain_decoder!(Int8Type),
+            Int16 => create_plain_decoder!(Int16Type),
+            Int32 => create_plain_decoder!(Int32Type),
+            Int64 => create_plain_decoder!(Int64Type),
+            UInt8 => create_plain_decoder!(UInt8Type),
+            UInt16 => create_plain_decoder!(UInt16Type),
+            UInt32 => create_plain_decoder!(UInt32Type),
+            UInt64 => create_plain_decoder!(UInt64Type),
+            Float16 => create_plain_decoder!(Float16Type),
+            Float32 => create_plain_decoder!(Float32Type),
+            Float64 => create_plain_decoder!(Float64Type),
+            _ => {
+                return Err(Error::new(
+                    ErrorKind::InvalidData,
+                    format!("Unsupport primitive type: {}", field.data_type()),
+                ))
+            }
+        };
+        decoder.decode().await
+    }
+
+    async fn read_binary_array(&self, field: &Field, batch_id: i32) -> Result<ArrayRef> {
+        let column = field.id;
+        let page_info = self.page_table.get(column, batch_id).ok_or(Error::new(
+            ErrorKind::InvalidData,
+            format!(
+                "field: {} batch {} does not exist in page table",
+                field, batch_id
+            ),
+        ))?;
+
+        use DataType::*;
+        let decoder: Box<dyn Decoder> = match field.data_type() {
```

Comment:
is LargeUtf8 and LargeBinary handled differently?

Reply:
LargeUtf8 / LargeBinary has i64 offset , which is different to i32 offset for utf8/binary